### PR TITLE
[KIP-7] Applied KIP-13

### PR DESCRIPTION
**Pull request by KimKyungup**
I wrote draft KIP-7 with KIP-13.

### CODE REVIEW

File: KIPs/kip-7-fungible_token.md [hunk 1]
```diff
@@ -27,9 +27,12 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
 ## Specification
 <!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Klaytn platforms (klaytn). -->
 This document derived heavily from Ethereum's [ERC-20 token standard](https://eips.ethereum.org/EIPS/eip-20) written by Fabian Vogelsteller and Vitalik Buterin.
-Comparing with ERC-20 token standard, this token standard has two differences:
-- includes three more optional functions (mint, burn, and burnFrom)
+Comparing with ERC-20 token standard, this token standard has some differences:
+- can include optional functions (e.g. mint, burn, pausable methods)
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
- can include optional functions (e.g. mintable, burnable and pausable methods)
```
How about matching form?
"mintable, burnable and pausable methods", "mint, burn and pause" or "mint, burn and pause methods"

File: KIPs/kip-7-fungible_token.md [hunk 2]
```diff
@@ -46,9 +49,46 @@ If the optional field is not marked, the function must be implemented.
...
+|Name|Prototype|
+|---|---|
+|[name](#name)|function name() public view returns (string)|
+|[symbol](#symbol)|function symbol() public view returns (string)|
+|[decimals](#decimals)|function decimals() public view returns (uint8)|
+
+
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with detailed methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for detailed methods.
+```solidity
+/// @title KIP-721 Non-Fungible Token Standard, optional detailed extension
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
/// @title KIP-7 Fungible Token Standard, optional detailed extension
```

File: KIPs/kip-7-fungible_token.md [hunk 3]
```diff
@@ -61,6 +101,20 @@ These methods are optional.
 |[addMinter](#addminter)|function addMinter(address _account) public |
 |[renounceMinter](#renounceminter)|function renounceMinter() public |
 
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with mint methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for mint methods.
+```solidity
+/// @title KIP-721 Non-Fungible Token Standard, optional mint extension
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
/// @title KIP-7 Fungible Token Standard, optional mint extension
```

File: KIPs/kip-7-fungible_token.md [hunk 4]
```diff
@@ -70,6 +124,19 @@ These methods are optional.
 |[burn](#burn)|function burn(uint256 _value) public |
 |[burnFrom](#burnfrom)|function burnFrom(address _from, uint256 _value) public |
 
+
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with burn methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for burn methods.
+```solidity
+/// @title KIP-721 Non-Fungible Token Standard, optional burn extension
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
/// @title KIP-7 Fungible Token Standard, optional burn extension
```

File: KIPs/kip-7-fungible_token.md [hunk 5]
```diff
@@ -83,6 +150,18 @@ These methods are optional.
 |[addPauser](#addminter)|function addPauser(address _account) public |
 |[renouncePauser](#renounceminter)|function renouncePauser() public |
 
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with Pause methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for Pause methods.
+```solidity
+/// @title KIP-721 Non-Fungible Token Standard, optional pause extension
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
/// @title KIP-7 Fungible Token Standard, optional pause extension
```

File: KIPs/kip-7-fungible_token.md [hunk 6]
```diff
@@ -83,6 +150,18 @@ These methods are optional.
...
+///  Note: the KIP-13 identifier for this interface is 0xXXXXXXXX.
+interface IKIP7Pause {
+    function supportsInterface(bytes4 interfaceID) external view returns (bool);    // KIP-13 interface
+
+    function transfer(address to, uint256 value) public whenNotPaused returns (bool);
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]: The example of IKIP7Pause methods should be changed.

Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: I fixed. :) Thanks.

File: KIPs/kip-7-fungible_token.md [hunk 7]
```diff
@@ -83,6 +150,18 @@ These methods are optional.
...
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract with Pause methods MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for Pause methods.
+```solidity
+/// @title KIP-7 Fungible Token Standard, optional pause extension
+///  Note: the KIP-13 identifier for this interface is 0xXXXXXXXX.
+interface IKIP7Pause {
+    function supportsInterface(bytes4 interfaceID) external view returns (bool);    // KIP-13 interface
+    
+    function paused() public view returns (bool)
+    function pause() public;
+    function unpause() public;
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]: Thank you for changing :) Can you add isPauser, addPauser and renouncePauser?

File: KIPs/kip-7-fungible_token.md [hunk 8]
```diff
@@ -46,9 +49,46 @@ If the optional field is not marked, the function must be implemented.
 |[transferFrom](#transferfrom)| |function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)|
 |[approve](#approve)| |function approve(address _spender, uint256 _value) public returns (bool success)|
 |[allowance](#allowance)| |function allowance(address _owner, address _spender) public view returns (uint256 remaining)|
-|[name](#name)|O|function name() public view returns (string)|
-|[symbol](#symbol)|O|function symbol() public view returns (string)|
-|[decimals](#decimals)|O|function decimals() public view returns (uint8)|
+
+Every [KIP-7](https://klaytn.github.io/kips/KIPs/kip-7-fungible_token) compliant contract MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) `supportsInterface` with the identifier `0xXXXXXXXX` for common methods.
+```solidity
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]: Could you please check out this? If you like it, please make this similar to it.

https://github.com/klaytn/kips/blob/dc5999a20b1746ff8b7dee49824c9b2adbfbfd5a/KIPs/kip-17-non_fungible_token.md#metadata-extension

Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: Ok. I will modify this like it.

Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: I updated. PTAL. :)

File: KIPs/kip-7-fungible_token.md [hunk 9]
```diff
@@ -26,399 +26,232 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
-
-|Name|Prototype|
-|---|---|
-|[mint](#mint)|function mint(address _to, uint256 _value) public returns (bool) |
-|[isMinter](#isminter)|function isMinter(address _account) public view returns (bool) |
-|[addMinter](#addminter)|function addMinter(address _account) public |
-|[renounceMinter](#renounceminter)|function renounceMinter() public |
+### Comparing with ERC-20
+Comparing with ERC-20 token standard, this token standard has some differences:
+- can include optional functions (e.g. mintable, burnable and pausable methods)
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
-  More optional functions are included (e.g. mintable, burnable and pausable methods)
```

File: KIPs/kip-7-fungible_token.md [hunk 10]
```diff
@@ -26,399 +26,232 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
+- Every token transfer/mint/burn MUST be tracked by event logs. This means that a Transfer event MUST be emitted for any action related to transfer/mint/burn.
+- MUST implement [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) interface for each method group.
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
- The [KIP-13](https://klaytn.github.io/kips/KIPs/kip-interface_query_standard) interface for each method group MUST be implemented.
```

File: KIPs/kip-7-fungible_token.md [hunk 11]
```diff
@@ -26,399 +26,232 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
 
-#### Burn Methods (Optional)
-The table below is a summary of methods related to destroying tokens.
-These methods are optional.
+### KIP-7 Identifiers
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
### KIP-13 Identifiers
```

File: KIPs/kip-7-fungible_token.md [hunk 12]
```diff
@@ -26,399 +26,232 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
+The below table shows KIP-13 identifiers for interfaces defined in this proposal.
 
-|Name|Prototype|
+|Interface|KIP-7 Identifier|
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
|Interface|KIP-13 Identifier|
```

File: KIPs/kip-7-fungible_token.md [hunk 13]
```diff
@@ -26,399 +26,232 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
-|[paused](#paused)|function paused() public view returns (bool) |
-|[pause](#pause)|function pause() public |
-|[unpause](#burnfrom)|function unpause() public |
-|[isPauser](#isminter)|function isPauser(address _account) public view returns (bool) |
-|[addPauser](#addminter)|function addPauser(address _account) public |
-|[renouncePauser](#renounceminter)|function renouncePauser() public |
-
-
-#### Events
+|IKIP7|0xXXXXXXXX (TBD)|
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]: When will this field be updated?

File: KIPs/kip-7-fungible_token.md [hunk 14]
```diff
@@ -26,399 +26,229 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
-
-Adds `_account` as a minter, and MUST fire the [MinterAdded event](#minteradded). The value of `_account` MUST be set with same account of `addMinter` method in the `MinterAdded` event.
-The function SHOULD `throw` if the `_account` is already a minter.
-The function SHOULD `throw` if the message sender is not a minter.
-
-OPTIONAL - This method can be used to improve usability,
-but interfaces and other contracts MUST NOT expect these values to be present.
+|[IKIP7](#kip7-interface)|0x36372b07|
+|[IKIP7Metadata](#metadata-extension)|0xa219a025|
+|[IKIP7Mint](#mint-extension)|0xeab83e20|
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]: How about matching the KIP-17 with the interface name?
KIP-17 is naming extension interface names in this way "~able".

Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
|[IKIP7Mint](#minting-extension)|0xeab83e20|
```

File: KIPs/kip-7-fungible_token.md [hunk 15]
```diff
@@ -26,399 +26,229 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
+|[IKIP7Burn](#burn-extension)|0x3b5a0bf8|
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
|[IKIP7Burn](#burning-extension)|0x3b5a0bf8|
```

File: KIPs/kip-7-fungible_token.md [hunk 16]
```diff
@@ -26,399 +26,229 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
+|[IKIP7Pause](#pause-extension)|0x4d5507ff|
```
Comment (jimni1222) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
|[IKIP7Pause](#pausing-extension)|0x4d5507ff|
```

File: KIPs/kip-7-fungible_token.md (hunk 13 above)
Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: I've updated already.

File: KIPs/kip-7-fungible_token.md (hunk 9 above)
Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: reflected.

File: KIPs/kip-7-fungible_token.md (hunk 10 above)
Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: reflected.

File: KIPs/kip-7-fungible_token.md (hunk 14 above)
Comment (KimKyungup) [reply KIPs/kip-7-fungible_token.md]: Reflected. Thanks.

File: KIPs/kip-7-fungible_token.md [hunk 17]
```diff
@@ -26,399 +26,229 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
-|[addPauser](#addminter)|function addPauser(address _account) public |
-|[renouncePauser](#renounceminter)|function renouncePauser() public |
+This document is heavily derived from [ERC-20](https://eips.ethereum.org/EIPS/eip-20) written by Fabian Vogelsteller and Vitalik Buterin.
 
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
 
-#### Events
+### Differences with ERC-20
+This section describes the differences between KIP-7 and ERC-20. 
+- More optional functions are included (e.g. mint, burn and pause extensions)
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
- More optional functions are included (e.g. mint, burn and pause extensions).
```

**Review by jimni1222 — APPROVED** [review]
LGTM :) Thank you for changing!

File: KIPs/kip-7-fungible_token.md [hunk 18]
```diff
@@ -26,399 +26,229 @@ A standard interface allows any token on Klaytn to be re-used by other applicati
...
+    function symbol() external view returns (string memory);
+
+    /// @dev Returns the number of decimals used to get its user representation.
+    /// For example, if `decimals` equals `2`, a balance of `505` tokens should
+    /// be displayed to a user as `5,05` (`505 / 10 ** 2`).
+    ///  Tokens usually opt for a value of 18, imitating the relationship between
+    /// Ether and Wei.
+    /// NOTE: This information is only used for _display_ purposes: it in
+    /// no way affects any of the arithmetic of the contract, including
+    /// {IKIP-7-balanceOf} and {IKIP-7-transfer}.
```
Comment (baerae-colin) [inline KIPs/kip-7-fungible_token.md]:
```suggestion
    /// `IKIP7.balanceOf` and `IKIP7.transfer`.
```

**Review by baerae-colin — APPROVED** [review]

**Pull request merged**